packages/integration-test/src/setup.js

```diff
@@ -0,0 +1,242 @@
+        const tokensToBeTransferred = this.config.numTestTokens.mul(this.scalingFactor);
+
+        // mint tokens
+        const publicOwnerBalance = await erc20.balanceOf(this.publicOwner);
+        if (publicOwnerBalance.lt(tokensToBeTransferred)) {
+            await erc20.mint(this.publicOwner, tokensToBeTransferred, this.opts);
+        }
+
+        // approve tokens
+        const publicOwnerApproval = await erc20.allowance(this.publicOwner, ace.address);
+        if (publicOwnerApproval.lt(tokensToBeTransferred)) {
+            await erc20.approve(ace.address, tokensToBeTransferred, this.opts);
+        }
+    }
+
+    /**
+     * @method getAddresses() - get the addresses for the desired contracts to deploy
+     */
+    getAddresses() {
+        const contractAddresses = getContractAddressesForNetwork(this.networkId);
+
+        if (contractAddresses === undefined) {
+            throw new AztecError(codes.NO_CONTRACT_ADDRESSES, {
+                message: 'No contract addresses exist for the given network ID in @aztec/contract-addresses',
+                networkId: this.networkId,
+            });
+        }
+        return contractAddresses;
+    }
+
+    /**
+     * @method getContractPromises - get the Truffle contracts for desired contracts, with the provider set
+     * @returns {Array} - returns an array of array. Each inner array is associated with one particular contract, the
+     * first index being the name of the contract and the second element the promise representing the truffle contract
+     */
+    getContractPromises() {
+        return this.contractsToDeploy.map((nameOfContract) => [nameOfContract, this.getContractPromise(nameOfContract)]);
+    }
+
+    /**
+     * @method getContractPromise - get an object whose key is the name of a contract, and the value is a function that
+     * returns the Truffle representation of the contract
+     * @param {string} nameOfContract
+     */
+    getContractPromise(nameOfContract) {
+        const extractedContractArtifact = contractArtifacts[nameOfContract];
+        if (extractedContractArtifact === undefined) {
+            throw new AztecError(codes.NO_CONTRACT_ARTIFACT, {
+                message: 'Requested contract does not exist in @aztec/contract-artifacts',
+                requestedContract: nameOfContract,
+            });
+        }
+
+        const extractedContractAddress = this.contractAddresses[nameOfContract];
+        if (extractedContractAddress === undefined) {
+            throw new AztecError(codes.NO_CONTRACT_ADDRESS, {
+                message: 'Requested contract does not have a deployed address in @aztec/contract-addresses',
+                requestedContract: nameOfContract,
+            });
+        }
+
+        const truffleRepresentation = this.getTruffleContractRepresentation(extractedContractArtifact);
+        return truffleRepresentation.at(extractedContractAddress); // this is going to return a promise
```

Comment:
Would be cleaner to return an `async` function which can be evaluated and awaited. Right now, unless am missing something with this code, you will start evaluating `truffleRepresentation.at(...)` as soon as this fn hits this line. You happen to be saving the `Promise` which this evaluation returns while waiting for the async fetch to actually return data, but this line already adds instructions to the event loop.

All in all, not catastrophic given this is only a fetch (although right now you are technically evaluating the promise multiple times, once when this line gets hit, the other when you await the promise this line returns). If this call had side effects, you could for example unwittingly be writing multiple times.

Reply:
Very good point, now return an `async` function